tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,76 @@
+#ifndef ROOT_RDF_RMERGEABLEVALUE
+#define ROOT_RDF_RMERGEABLEVALUE
+
+#include <functional>
+#include <memory>
+#include "RtypesCore.h"
+
+namespace ROOT {
+namespace RDF {
+
+template <typename T>
+class RMergeableValue {
+protected:
+   T fValue;
+
+public:
+   RMergeableValue() = delete;
+   RMergeableValue(T &value) : fValue{value} {}
+   virtual ~RMergeableValue() = default;
+   virtual void Merge(const std::unique_ptr<RMergeableValue<T>> &) = 0;
+   virtual ULong64_t GetCounts() { return 0; }
+   virtual T &GetValue() { return fValue; }
+};
+
+// Specialization for generic action
+template <typename T>
+class RMergeableGenericAction final : public RMergeableValue<T> {
+   using mergefun_t = std::function<T(T &, T &)>;
+   const mergefun_t fMergeFun;
+
+public:
+   RMergeableGenericAction() = delete;
+   RMergeableGenericAction(T &value, const mergefun_t &mergefun) : RMergeableValue<T>(value), fMergeFun{mergefun} {}
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) override
+   {
+      auto &othervalue = other->GetValue();
```

Comment:
It can't be generally since the intent of `RMergeableValue` is to change the value of the `fValue` data member and then provide it to the user once merged via `GetValue`